src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java

```diff
@@ -0,0 +1,90 @@
+package org.broadinstitute.hellbender.engine.spark.datasources;
+
+import htsjdk.samtools.*;
+import org.apache.commons.collections4.iterators.IteratorIterable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadCoordinateComparator;
+import org.seqdoop.hadoop_bam.KeyIgnoringBAMOutputFormat;
+import org.seqdoop.hadoop_bam.SAMRecordWritable;
+import scala.Tuple2;
+
+import java.io.IOException;
+
+/**
+ * ReadsSparkSink writes GATKReads to a file. This code lifts from the HadoopGenomics/Hadoop-BAM
+ * read writing code as well as from bigdatagenomics/adam.
+ */
+public class ReadsSparkSink {
+
+    // We need an output format for saveAsNewAPIHadoopFile.
+    public static class MyOutputFormat extends KeyIgnoringBAMOutputFormat<NullWritable> {
+        public static SAMFileHeader bamHeader = null;
+
+        public static void setHeader(final SAMFileHeader header) {
+            bamHeader = header;
+        }
+
+        @Override
+        public RecordWriter<NullWritable, SAMRecordWritable> getRecordWriter(TaskAttemptContext ctx) throws IOException {
+            setSAMHeader(bamHeader);
+            return super.getRecordWriter(ctx);
+        }
+    }
+
+    /**
+     * writeParallelReads writes rddReads to outputFile with header as the file header.
+     * @param ctx the JavaSparkContext to write.
+     * @param outputFile path to the output bam.
+     * @param rddReads reads to write.
+     * @param header the header to put at the top of the file.s
+     */
+    public static void writeParallelReads(
+            final JavaSparkContext ctx, final String outputFile, final JavaRDD<GATKRead> rddReads,
+            final SAMFileHeader header) throws IOException {
+        // Set the header on the main thread.
+        MyOutputFormat.setHeader(header);
```

Comment:
Using configuration is more general (e.g. it works for MapReduce and other frameworks, which is why it is used in Hadoop-BAM), and gives you more control over the serialization.

In this context, where Spark is being used explicitly, broadcasting is better as it's more efficient (as you pointed out). So please ignore my earlier comment about using configuration :)